### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by s0nicboOm**
MAJOR:
Introduce metrics to the udfs.

MINOR:
1. Fix logs for insufficient data
2. Add logs in the scenario preproc model is absent and we forward the message to trainer directly.

### CODE REVIEW

File: numalogic/udfs/__main__.py [hunk 1]
```diff
@@ -37,6 +38,10 @@ def start_server() -> None:
 
     LOGGER.info("Running %s on %s server", step, server_type)
 
+    # TODO: Add metrics server port to config
+    # Start the metrics server at port 8490
+    start_metrics_server(8490)
```
Comment (ab93) [inline numalogic/udfs/__main__.py:43]: Get the port from a env var and let's default to 8490

File: numalogic/udfs/inference.py [hunk 2]
```diff
@@ -95,6 +103,10 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
         # Construct payload object
         payload = StreamPayload(**orjson.loads(datum.value))
 
+        MSG_IN_COUNTER.increment_counter(
+            self.__class__.__name__, ":".join(payload.composite_keys), payload.config_id
```
Comment (ab93) [inline numalogic/udfs/inference.py:107]: Let's add a `_vtx = "inference"` or something of that sort as a class variable.

File: numalogic/udfs/inference.py [hunk 3]
```diff
@@ -161,6 +196,15 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
             payload.uuid,
             time.perf_counter() - _start_time,
         )
+        MODEL_STATUS_COUNTER.increment_counter(
+            payload.status.value,
+            self.__class__.__name__,
+            ":".join(payload.composite_keys),
+            payload.config_id,
+        )
```
Comment (ab93) [inline numalogic/udfs/inference.py:204]: Can you create a method that takes in these vars along with the right Counter object that does this?

**Review by ab93 — COMMENTED** [review]
Few comments

File: numalogic/udfs/inference.py (hunk 2 above)
Comment (s0nicboOm) [reply numalogic/udfs/inference.py:107]: Sure. Let me introduce this in base class itself

File: numalogic/monitoring/metrics.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    __slots__ = ("name", "description", "labels", "counter")
+
+    def __init__(self, name: str, description: str, labels: list[str]) -> None:
+        self.name = name
+        self.description = description
+        self.labels = labels
+        self.counter = Counter(name, description, labels)
+
+    def increment_counter(self, *args, amount: float = 1) -> None:
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:27]:
```suggestion
    def increment_counter(self, *labels, amount: float = 1) -> None:
```

Comment (ab93) [reply numalogic/monitoring/metrics.py:27]:
```suggestion
    def increment_counter(self, *labels, amount: int = 1) -> None:
```

File: numalogic/udfs/inference.py (hunk 3 above)
Comment (s0nicboOm) [reply numalogic/udfs/inference.py:204]: Makes sense

File: numalogic/monitoring/metrics.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import logging
+from typing import Optional
+
+from prometheus_client import Counter, Info, Summary
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BaseMetricClass:
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:9]:
```suggestion
class _BaseMetric:
```
Why `Class` in the name of a class? :)

File: numalogic/monitoring/metrics.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+class PromCounterMetric(BaseMetricClass):
+    """Class is used to create a counter object and increment it."""
+
+    __slots__ = "counter"
+
+    def __init__(self, name: str, description: str, labels: list[str]) -> None:
+        super().__init__(name, description, labels)
+        self.counter = Counter(name, description, labels)
+
+    def increment_counter(self, *args, amount: int = 1) -> None:
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:36]:
```suggestion
    def increment_counter(self, *label_values, amount: int = 1) -> None:
```

File: numalogic/monitoring/metrics.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    Args:
+        name: Name of the metric
+        description: Description of the metric
+        labels: List of labels
+    """
+
+    __slots__ = ("name", "description", "labels")
+
+    def __init__(self, name: str, description: str, labels: Optional[list[str]]) -> None:
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:21]:
```suggestion
    def __init__(self, name: str, description: str, label_keys: Optional[list[str]]) -> None:
```

File: numalogic/monitoring/metrics.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        if len(args) != len(self.labels):
+            raise ValueError(f"Labels mismatch with the definition: {self.labels}")
+        self.counter.labels(*args).inc(amount=amount)
+
+
+class PromInfoMetric(BaseMetricClass):
+    """Class is used to create an info object and increment it."""
+
+    __slots__ = ("name", "description", "labels", "info")
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:52]: You don't need to redefine the base class slots again

File: numalogic/monitoring/metrics.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(self, name: str, description: str, labels: Optional[list[str]]) -> None:
+        super().__init__(name, description, labels)
+        self.info = Info(name, description, labels)
+
+    def add_info(
+        self,
+        *args,
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:60]: same as above

File: numalogic/monitoring/metrics.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        if len(args) != len(self.labels):
+            raise ValueError(f"Labels mismatch with the definition: {self.labels}")
+        self.info.labels(*args).info(data)
+
+
+class PromSummaryMetric(BaseMetricClass):
+    """Class is used to create a histogram object and increment it."""
+
+    __slots__ = ("name", "description", "labels", "summary")
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:78]: same as before

File: numalogic/udfs/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+|        Metric Name        |   Type    |                  Labels                  |                   Description                   |
```
Comment (ab93) [inline numalogic/udfs/README.md:1]: Very nice!

File: numalogic/udfs/_metrics.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    "RUNTIME_ERROR_COUNTER",
+    "MSG_IN_COUNTER",
+    "MSG_PROCESSED_COUNTER",
+    "MODEL_INFO",
+    "UDF_TIME",
+    "FETCH_TIME_SUMMARY",
+    "DATAFRAME_SHAPE_SUMMARY",
+    "FETCH_EXCEPTION_COUNTER",
+    "INF_SUMMARY",
+]
```
Comment (ab93) [inline numalogic/udfs/_metrics.py:26]: Any reason why we are adding this here?

File: numalogic/udfs/preprocess.py [hunk 13]
```diff
@@ -151,12 +189,34 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
             )
             # TODO check again what error is causing this and if retraining is required
             payload = replace(payload, status=Status.RUNTIME_ERROR, header=Header.TRAIN_REQUEST)
+            _increment_counter(
+                counter=MODEL_STATUS_COUNTER,
+                labels=(
+                    payload.status.value,
+                    self._vtx,
+                    ":".join(payload.composite_keys),
+                    payload.config_id,
```
Comment (ab93) [inline numalogic/udfs/preprocess.py:198]:
```suggestion
                    payload.status.value,
                    *_label_values,
```
And define the `_label_values` as a local variable in the exec(). Seems we are adding this same subset many times.

**Review by ab93 — CHANGES_REQUESTED** [review]

File: numalogic/udfs/preprocess.py (hunk 13 above)
Comment (s0nicboOm) [reply numalogic/udfs/preprocess.py:198]: Yes a very good idea! Let me make that

File: numalogic/monitoring/metrics.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
+import logging
+from typing import Optional
+
+from prometheus_client import Counter, Info, Summary
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class _BaseMetric:
+    __slots__ = "name", "description", "label_keys"
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:10]:
```suggestion
    __slots__ = ("name", "description", "label_keys")
```

File: numalogic/udfs/_metrics.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+# helper functions
+
+
+def _increment_counter(counter: PromCounterMetric, labels: tuple, amount: int = 1) -> None:
+    """
+    Utility function is used to increment the counter.
+
+    Args:
+        counter: Counter object
+        *args: List of labels
```
Comment (ab93) [inline numalogic/udfs/_metrics.py:115]: update docstring

File: numalogic/udfs/_metrics.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    "numalogic_udf_time_histogram",
+    "Histogram for udf processing time",
+    buckets=buckets,
+)
+
+
+# helper functions
+
+
+def _increment_counter(counter: PromCounterMetric, labels: tuple, amount: int = 1) -> None:
```
Comment (ab93) [inline numalogic/udfs/_metrics.py:109]: why not *labels?

**Review by ab93 — APPROVED** [review]
Approved, with a few comments

File: numalogic/udfs/_metrics.py (hunk 16 above)
Comment (s0nicboOm) [reply numalogic/udfs/_metrics.py:109]: Here, you don't need *labels right? This taking in a tuple makes more explainability for the users. What do you think?

**Pull request merged**